File: src/main/java/id/walt/ssikitexamples/CustomData.java
package id.walt.ssikitexamples;

import id.walt.credentials.w3c.builder.W3CCredentialBuilder;
import id.walt.credentials.w3c.templates.VcTemplate;
import id.walt.credentials.w3c.templates.VcTemplateService;
import id.walt.crypto.KeyAlgorithm;
import id.walt.model.DidMethod;
import id.walt.servicematrix.ServiceMatrix;
import id.walt.services.did.DidService;
import id.walt.services.key.KeyService;
import id.walt.signatory.Ecosystem;
import id.walt.signatory.ProofConfig;
import id.walt.signatory.ProofType;
import id.walt.signatory.Signatory;
import id.walt.signatory.dataproviders.MergingDataProvider;

import java.util.AbstractMap.SimpleEntry;
import java.util.List;
import java.util.Map;

public class CustomData {

    public static void main(String[] args) {
        new CustomData().run();
    }

    public void run() {
        // Load walt.id SSI-Kit services from "$workingDirectory/service-matrix.properties"
        new ServiceMatrix("service-matrix.properties");

        // Define used services
        var signatory = Signatory.Companion.getService();
        var keyService = KeyService.Companion.getService();

        /* Use services... */
        // generate key pairs for holder, issuer
        var holderKey = keyService.generate(KeyAlgorithm.EdDSA_Ed25519);
        var issuerKey = keyService.generate(KeyAlgorithm.EdDSA_Ed25519);

        // create dids, using did:key
        var holderDid = DidService.INSTANCE.create(DidMethod.key, holderKey.getId(), null);
        var issuerDid = DidService.INSTANCE.create(DidMethod.key, issuerKey.getId(), null);

        // List registered VC templates
        List<VcTemplate> vcTemplates = signatory.listTemplates();
        vcTemplates.forEach(template -> {
            System.out.println(vcTemplates.indexOf(template) + ": " + template);
        });

        // Create VC template
        var verifiableDiplomaTemplate = VcTemplateService.Companion.getService().getTemplate("VerifiableDiploma", true, "");
        System.out.println("Default Verifiable Diploma - " + verifiableDiplomaTemplate);

        // Prepare desired custom data that should replace the default template data
        var data = Map.ofEntries(
                credentialSubjectEntry(new SimpleEntry("givenNames", "Yves"), new SimpleEntry("familyName", "SMITH"), new SimpleEntry("dateOfBirth", "2000-02-04"))
        );

        // Custom VC template
        var verifiableDiploma = new MergingDataProvider(data).populate(
                new W3CCredentialBuilder(),
                new ProofConfig(issuerDid, holderDid, null, null, ProofType.LD_PROOF, null, null,
                        null, null, null, null, null, null, null,null, Ecosystem.DEFAULT)
        ).build();
        System.out.println("Verifiable Diploma with custom data - " + verifiableDiploma.encode());

    }

    private SimpleEntry<String, Map<String, String>> credentialSubjectEntry(SimpleEntry<String, String>... pairs) {
        return new SimpleEntry("credentialSubject", Map.ofEntries(pairs));
    }
}

File: src/main/kotlin/id/walt/ssikitexamples/CustomData.kt
package id.walt.ssikitexamples

import id.walt.auditor.Auditor
import id.walt.auditor.JsonSchemaPolicy
import id.walt.auditor.SignaturePolicy
import id.walt.auditor.VerificationPolicy
import id.walt.custodian.Custodian
import id.walt.servicematrix.ServiceMatrix
import id.walt.signatory.DataProviderRegistry
import id.walt.signatory.ProofConfig
import id.walt.signatory.ProofType
import id.walt.signatory.Signatory
import id.walt.vclib.model.VerifiableCredential
import id.walt.vclib.vclist.VerifiableId
import id.walt.vclib.vclist.VerifiablePresentation

class MyCustomPolicy : VerificationPolicy {
    override val description: String
        get() = "A custom verification policy"

    override fun verify(vc: VerifiableCredential): Boolean {
        if (vc is VerifiableId) {
            val idData = MockedIdDatabase.get(vc.credentialSubject!!.id!!)
            if (idData != null) {
                return idData.familyName == vc.credentialSubject?.familyName && idData.firstName == vc.credentialSubject?.firstName
            }
        } else if (vc is VerifiablePresentation) {
            // This custom policy does not verify the VerifiablePresentation
            return true
        }
        return false
    }
}

val signatory = Signatory.getService()
val custodian = Custodian.getService()

fun main() {
    // Load Walt.ID SSI-Kit services from "$workingDirectory/service-matrix.properties"
    ServiceMatrix("service-matrix.properties")

    // Create VCs to verify:
    val idIter = MockedIdDatabase.mockedIds.keys.iterator()
    val holder = idIter.next()
    val issuer = idIter.next()

    // Register custom data provider
    DataProviderRegistry.register(VerifiableId::class, CustomIdDataProvider())

    // Issue VC in JSON-LD and JWT format (for show-casing both formats)
    val vcJson =
        signatory.issue("VerifiableId", ProofConfig(issuerDid = issuer, subjectDid = holder, proofType = ProofType.LD_PROOF))
    val vcJwt = signatory.issue("VerifiableId", ProofConfig(issuerDid = issuer, subjectDid = holder, proofType = ProofType.JWT))

    // Present VC in JSON-LD and JWT format (for show-casing both formats)
    val vpJson = custodian.createPresentation(listOf(vcJson), holder)
    val vpJwt = custodian.createPresentation(listOf(vcJwt), holder)

    // Verify VPs, using Signature, JsonSchema and a custom policy
    val resJson = Auditor.verify(vpJson, listOf(SignaturePolicy(), JsonSchemaPolicy(), MyCustomPolicy()))
    val resJwt = Auditor.verify(vpJwt, listOf(SignaturePolicy(), JsonSchemaPolicy(), MyCustomPolicy()))

    println("JSON verification result: ${resJson.overallStatus}")
    println("JWT verification result: ${resJwt.overallStatus}")
}
